File: numasec/core/planner.py
"""Deterministic Planner — PTES 5-phase based, NOT LLM-driven."""

from __future__ import annotations

import hashlib
import logging
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from numasec.models.enums import PTESPhase
from numasec.models.plan import AttackPhase, AttackPlan, AttackStep

if TYPE_CHECKING:
    from numasec.models.target import TargetProfile


@dataclass
class ReplanSignal:
    """Signal from the Perceptor to the Planner (replanning triggers)."""

    type: str = "continue"
    confidence: float = 0.5
    data: dict = field(default_factory=dict)
    finding: object | None = None
    replan_reason: str | None = None

logger = logging.getLogger("numasec.core.planner")


@dataclass
class Task:
    """Single task within a plan."""

    id: str
    description: str
    tool: str = ""
    category: str = ""
    params: dict = None
    expected_output: str = ""

    def __post_init__(self):
        if self.params is None:
            self.params = {}


class DeterministicPlanner:
    """Deterministic planner following PTES methodology.

    The LLM does NOT decide strategy — only the Executor uses LLM for
    tactical decisions. Re-planning occurs only on Perceptor triggers.

    Based on CHECKMATE paper (arXiv:2512.11143).
    """

    SCOPE_TIMEOUTS = {
        "quick": 3,
        "standard": 10,
        "deep": 30,
    }

    def create_plan(self, target: TargetProfile, scope: str = "quick") -> AttackPlan:
        """Generate a deterministic attack plan based on target and scope."""
        plan = AttackPlan(target=target.target, scope=scope)

        # Phase 1: Reconnaissance (always)
        recon_steps = [
            AttackStep(id="recon_port_scan", description="Port scan target", tool="scan_engine"),
            AttackStep(id="recon_tech_fp", description="Technology fingerprint", tool="http_engine"),
        ]
        if scope in ("standard", "deep"):
            recon_steps.append(AttackStep(id="recon_dir_fuzz", description="Directory fuzzing", tool="http_engine"))
        plan.phases.append(
            AttackPhase(
                phase=PTESPhase.RECON,
                steps=recon_steps,
                parallelizable=True,
                timeout_minutes=self.SCOPE_TIMEOUTS.get(scope, 3),
            )
        )

        # Phase 2: Service Mapping
        plan.phases.append(
            AttackPhase(
                phase=PTESPhase.MAPPING,
                steps=[
                    AttackStep(id="map_services", description="Map detected services", tool="scan_engine"),
                    AttackStep(id="map_endpoints", description="Map web endpoints", tool="http_engine"),
                ],
                timeout_minutes=5,
            )
        )

        # Phase 3: Vulnerability Testing
        vuln_steps = self._select_vuln_tests(target)
        plan.phases.append(
            AttackPhase(
                phase=PTESPhase.VULNERABILITY,
                steps=vuln_steps,
                parallelizable=True,
                timeout_minutes=self.SCOPE_TIMEOUTS.get(scope, 10),
            )
        )

        # Phase 4: Exploitation Validation
        if scope in ("standard", "deep"):
            plan.phases.append(
                AttackPhase(
                    phase=PTESPhase.EXPLOITATION,
                    steps=[
                        AttackStep(
                            id="exploit_validate", description="Validate exploitable findings", tool="http_engine"
                        ),
                    ],
                    timeout_minutes=15,
                )
            )

        # Phase 5: Reporting (always)
        plan.phases.append(
            AttackPhase(
                phase=PTESPhase.REPORTING,
                steps=[
                    AttackStep(id="report_gen", description="Generate assessment report", tool="reporter"),
                ],
                timeout_minutes=2,
            )
        )

        logger.info(
            "Created plan: %d phases, %d total steps, scope=%s",
            len(plan.phases),
            sum(len(p.steps) for p in plan.phases),
            scope,
        )
        return plan

    def replan(
        self,
        current_plan: AttackPlan,
        signal: ReplanSignal,
        discovered_nodes: frozenset[str] = frozenset(),
    ) -> AttackPlan:
        """Re-plan on Perceptor triggers (WAF detected, escalation found, etc.).

        Args:
            current_plan: The existing plan to mutate.
            signal: The Perceptor signal that triggered replanning.
            discovered_nodes: Attack graph node IDs already discovered.
                Used to augment VULNERABILITY phase steps based on confirmed
                capabilities (RF4 — AttackGraph → Planner wiring).
        """
        logger.info("Replanning triggered by: %s", signal.type)

        if signal.type == "waf_detected":
            # Add WAF evasion steps to vulnerability phase
            for phase in current_plan.phases:
                if phase.phase == PTESPhase.VULNERABILITY:
                    phase.steps.append(
                        AttackStep(
                            id="waf_evasion",
                            description="Attempt WAF bypass techniques (encoding, case alternation, comment injection)",
                            tool="http_engine",
                        )
                    )
                    break

        elif signal.type == "escalation_found":
            chain_name = signal.data.get("chain", "")
            for phase in current_plan.phases:
                if phase.phase == PTESPhase.EXPLOITATION:
                    phase.steps.append(
                        AttackStep(
                            id=f"escalate_{chain_name}",
                            description=f"Escalation chain: {chain_name}",
                            tool="http_engine",
                        )
                    )
                    break

        elif signal.type == "auth_obtained":
            # Credentials discovered -- insert post-auth testing tiers
            token_type = signal.data.get("token_type", "bearer")
            for phase in current_plan.phases:
                if phase.phase == PTESPhase.VULNERABILITY:
                    step_id = self._step_id("auth_retest", "post_auth")
                    if not any(s.id == step_id for s in phase.steps):
                        phase.steps.append(
                            AttackStep(
                                id=step_id,
                                description=(
                                    f"Post-auth testing with {token_type} token: "
                                    "IDOR, privilege escalation, business logic, stored XSS"
                                ),
                                tool="get_auth_retest_plan",
                            )
                        )
                    break

        elif signal.type == "rate_limited":
            # Reduce scan aggressiveness
            logger.warning("Rate limiting detected -- adding throttle steps")
            for phase in current_plan.phases:
                if phase.phase == PTESPhase.VULNERABILITY:
                    if not any(s.id == "throttle_mode" for s in phase.steps):
                        phase.steps.insert(
                            0,
                            AttackStep(
                                id="throttle_mode",
                                description="Rate limiting detected: increase delay between requests to 2s, reduce concurrency to 1",
                                tool="http_engine",
                            ),
                        )
                    break

        elif signal.type == "technology_identified":
            # Tech-specific scanner prioritisation
            tech = signal.data.get("technology", "").lower()
            tech_steps: list[tuple[str, str, str]] = []

            if "node" in tech or "express" in tech:
                tech_steps.append(("nosql_deep", "Deep NoSQL injection (Node.js target)", "nosql_test"))
                tech_steps.append(("proto_poll", "Prototype pollution check (Node.js)", "command_injection_test"))
            elif "php" in tech:
                tech_steps.append(("lfi_deep", "Deep LFI with PHP wrappers (expect://, data://)", "lfi_test"))
            elif "java" in tech or "spring" in tech:
                tech_steps.append(("deser_check", "Deserialization check (Java target)", "vuln_scan"))
            elif "django" in tech or "flask" in tech or "python" in tech:
                tech_steps.append(("ssti_deep", "Deep SSTI Jinja2 check (Python target)", "ssti_test"))

            for step_id, desc, tool in tech_steps:
                for phase in current_plan.phases:
                    if phase.phase == PTESPhase.VULNERABILITY:
                        full_id = self._step_id(step_id, tech)
                        if not any(s.id == full_id for s in phase.steps):
                            phase.steps.append(AttackStep(id=full_id, description=desc, tool=tool))
                        break

        elif signal.type == "large_surface":
            # Auto-escalate scope when many endpoints discovered
            endpoint_count = signal.data.get("endpoint_count", 0)
            if endpoint_count > 100:
                logger.info("Large attack surface (%d endpoints) -- escalating scope", endpoint_count)
                current_plan.scope = "deep"

        elif signal.type == "api_app_detected":
            # Prioritise JSON/API-specific scanners
            for phase in current_plan.phases:
                if phase.phase == PTESPhase.VULNERABILITY:
                    step_id = self._step_id("api_scan", "api")
                    if not any(s.id == step_id for s in phase.steps):
                        phase.steps.append(
                            AttackStep(
                                id=step_id,
                                description="API-specific testing: IDOR/BOLA, mass assignment, auth bypass",
                                tool="idor_test",
                            )
                        )
                    break

        elif signal.type == "spa_detected":
            # Switch to browser-first scanning
            for phase in current_plan.phases:
                if phase.phase == PTESPhase.MAPPING:
                    step_id = self._step_id("spa_crawl", "browser")
                    if not any(s.id == step_id for s in phase.steps):
                        phase.steps.append(
                            AttackStep(
                                id=step_id,
                                description="SPA detected: browser crawl for fragment routes and JS bundle analysis",
                                tool="browser_crawl_site",
                            )
                        )
                    break

        # RF4: Augment VULNERABILITY phase based on discovered graph nodes.
        if discovered_nodes:
            self._augment_vuln_phase_from_graph(current_plan, discovered_nodes)

        return current_plan

    def _augment_vuln_phase_from_graph(
        self,
        plan: AttackPlan,
        discovered_nodes: frozenset[str],
    ) -> None:
        """Inject additional vulnerability tests inferred from the attack graph.

        Called during replan when the graph has new discovered nodes that the
        initial deterministic plan did not account for.
        """
        vuln_phase = next((p for p in plan.phases if p.phase == PTESPhase.VULNERABILITY), None)
        if vuln_phase is None:
            return

        existing_ids = {s.id for s in vuln_phase.steps}

        # Directory enumeration found real content → add info-disclosure check
        if "recon_dirs" in discovered_nodes and "vuln_info_disclosure_check" not in existing_ids:
            vuln_phase.steps.append(
                AttackStep(
                    id="vuln_info_disclosure_check",
                    description="Check discovered paths for sensitive information disclosure",
                    tool="http_engine",
                )
            )
            logger.debug("Graph-guided: added vuln_info_disclosure_check (recon_dirs discovered)")

        # SQLi confirmed → add deep sqlmap escalation if not already present
        if "vuln_sqli" in discovered_nodes and "vuln_sqli_deep" not in existing_ids:
            vuln_phase.steps.append(
                AttackStep(
                    id="vuln_sqli_deep",
                    description="Deep SQL injection escalation via sqlmap",
                    tool="sqlmap",
                )
            )
            logger.debug("Graph-guided: added vuln_sqli_deep (vuln_sqli discovered)")

        # SSRF indicators found → add internal network probe
        if "vuln_ssrf" in discovered_nodes and "vuln_ssrf_probe" not in existing_ids:
            vuln_phase.steps.append(
                AttackStep(
                    id="vuln_ssrf_probe",
                    description="Probe internal services via confirmed SSRF",
                    tool="http_engine",
                )
            )
            logger.debug("Graph-guided: added vuln_ssrf_probe (vuln_ssrf discovered)")

        # LFI confirmed → add log-poisoning / wrapper escalation
        if "vuln_lfi" in discovered_nodes and "vuln_lfi_escalate" not in existing_ids:
            vuln_phase.steps.append(
                AttackStep(
                    id="vuln_lfi_escalate",
                    description="Attempt LFI to RCE via log poisoning or PHP wrappers",
                    tool="http_engine",
                )
            )
            logger.debug("Graph-guided: added vuln_lfi_escalate (vuln_lfi discovered)")

    def _select_vuln_tests(self, target: TargetProfile) -> list[AttackStep]:
        """Select vulnerability tests based on detected technologies."""
        steps: list[AttackStep] = []

        # Always run header checks
        steps.append(
            AttackStep(
                id="vuln_headers",
                description="Security header analysis",
                tool="http_engine",
            )
        )

        # Check for web services
        has_web = any(p.service in ("http", "https", "http-proxy") for p in target.ports)

        if has_web:
            steps.extend(
                [
                    AttackStep(id="vuln_sqli", description="SQL injection testing", tool="http_engine"),
                    AttackStep(id="vuln_xss", description="XSS testing", tool="http_engine"),
                    AttackStep(id="vuln_auth", description="Authentication testing", tool="http_engine"),
                ]
            )

        # Technology-specific tests
        tech_names = [t.name.lower() for t in target.technologies]

        if any("php" in t for t in tech_names):
            steps.append(
                AttackStep(
                    id="vuln_lfi",
                    description="LFI/RFI testing (PHP)",
                    tool="http_engine",
                )
            )

        if any(t in ("mysql", "postgresql", "mssql") for t in tech_names):
            steps.append(
                AttackStep(
                    id="vuln_sqli_deep",
                    description="Deep SQLi testing",
                    tool="sqlmap",
                )
            )

        if any("wordpress" in t for t in tech_names):
            steps.append(
                AttackStep(
                    id="vuln_wp",
                    description="WordPress vulnerability scan",
                    tool="nuclei",
                )
            )

        # If no tech detected, add generic nuclei scan
        if not tech_names:
            steps.append(
                AttackStep(
                    id="vuln_nuclei",
                    description="Generic vulnerability scan",
                    tool="nuclei",
                )
            )

        return steps

    # ------------------------------------------------------------------
    # Deterministic task generation (post-MAPPING expansion)
    # ------------------------------------------------------------------

    # Maps scanner tools to endpoint matching predicates.
    # Each predicate receives an Endpoint and returns True if the tool
    # should be run against that endpoint.
    _SCANNER_RULES: dict[str, str] = {
        "sqli_test": "query_or_post",
        "xss_test": "query_or_post",
        "nosql_test": "json_body",
        "csrf_test": "post_method",
        "idor_test": "numeric_path",
        "xxe_test": "upload_or_xml",
        "ssrf_test": "url_param",
        "ssti_test": "query_or_post",
    }

    # Tools that should always run once on the base URL.
    _ALWAYS_RUN_TOOLS: list[str] = [
        "auth_test",
        "cors_test",
        "host_header_test",
        "open_redirect_test",
        "access_control_test",
    ]

    # --- Helpers ---

    @staticmethod
    def _step_id(tool: str, url: str) -> str:
        """Generate a deterministic step ID from tool + URL."""
        url_hash = hashlib.sha256(url.encode()).hexdigest()[:8]
        return f"auto_{tool}_{url_hash}"

    @staticmethod
    def _endpoint_matches_rule(ep: Any, rule: str) -> bool:
        """Check if an endpoint matches a scanner selection rule."""
        url = getattr(ep, "url", "")
        method = getattr(ep, "method", "GET").upper()
        content_type = getattr(ep, "content_type", "").lower()
        parameters = getattr(ep, "parameters", [])

        if rule == "query_or_post":
            return bool(parameters) or method == "POST" or "?" in url
        elif rule == "json_body":
            return "json" in content_type or method == "POST"
        elif rule == "post_method":
            return method == "POST"
        elif rule == "numeric_path":
            return any(seg.isdigit() for seg in url.split("/"))
        elif rule == "upload_or_xml":
            return "upload" in url.lower() or "xml" in content_type or "file" in url.lower()
        elif rule == "url_param":
            lower_url = url.lower()
            return any(p in lower_url for p in ("url=", "image", "file", "upload", "img", "src"))
        return False

File: agent/packages/numasec/src/security/planner/planner.ts
/**
 * Deterministic Planner — PTES 5-phase methodology.
 *
 * Based on CHECKMATE paper (arxiv 2512.11143). No LLM calls.
 * Generates a structured attack plan based on target profile.
 */

export type PTESPhase = "recon" | "service_mapping" | "vuln_testing" | "exploitation" | "reporting"
export type Scope = "quick" | "standard" | "deep"

export interface AttackStep {
  tool: string
  description: string
  parameters: Record<string, any>
  priority: number // 1=must, 2=should, 3=nice
  condition?: string // when to include this step
}

export interface AttackPhase {
  phase: PTESPhase
  steps: AttackStep[]
}

export interface AttackPlan {
  target: string
  scope: Scope
  phases: AttackPhase[]
  timeout: number // seconds
  notes: string[]
}

interface TargetProfile {
  url: string
  technologies?: string[]
  openPorts?: number[]
  hasGraphql?: boolean
  hasApi?: boolean
  hasAuth?: boolean
  hasUpload?: boolean
  hasForms?: boolean
}

export function generatePlan(target: TargetProfile, scope: Scope = "standard"): AttackPlan {
  const timeouts: Record<Scope, number> = { quick: 180, standard: 600, deep: 1800 }
  const notes: string[] = []
  const phases: AttackPhase[] = []

  // Phase 1: Reconnaissance
  const reconSteps: AttackStep[] = [
    { tool: "recon", description: "Port scan, service detection, tech fingerprint", parameters: { target: target.url }, priority: 1 },
    { tool: "crawl", description: "Discover URLs, forms, technologies", parameters: { url: target.url }, priority: 1 },
  ]

  if (scope !== "quick") {
    reconSteps.push({
      tool: "dir_fuzz",
      description: "Brute-force hidden directories",
      parameters: { url: target.url },
      priority: 2,
    })
    reconSteps.push({
      tool: "js_analyze",
      description: "Analyze JavaScript for endpoints and secrets",
      parameters: { url: target.url },
      priority: 2,
    })
  }

  phases.push({ phase: "recon", steps: reconSteps })

  // Phase 2: Service Mapping (implicit in recon results)
  // No separate phase needed — recon + crawl cover this.

  // Phase 3: Vulnerability Testing
  const vulnSteps: AttackStep[] = []

  // Always test for injection
  vulnSteps.push({
    tool: "injection_test",
    description: "Test discovered parameters for SQL/command/template injection",
    parameters: { url: target.url },
    priority: 1,
  })

  // Always test for XSS
  vulnSteps.push({
    tool: "xss_test",
    description: "Test parameters for reflected XSS",
    parameters: { url: target.url },
    priority: 1,
  })

  // SSRF if URL parameters detected
  vulnSteps.push({
    tool: "ssrf_test",
    description: "Test URL-accepting parameters for SSRF",
    parameters: { url: target.url },
    priority: 2,
    condition: "url_params_found",
  })

  // Auth testing if login/auth endpoints
  if (target.hasAuth !== false) {
    vulnSteps.push({
      tool: "auth_test",
      description: "Test authentication: default creds, JWT, session management",
      parameters: { url: target.url },
      priority: 1,
    })
  }

  // Access control
  vulnSteps.push({
    tool: "access_control_test",
    description: "Test IDOR, CSRF, CORS, mass assignment",
    parameters: { url: target.url, test_type: "all" },
    priority: 1,
  })

  // GraphQL-specific
  if (target.hasGraphql) {
    vulnSteps.push({
      tool: "graphql_test",
      description: "Test GraphQL: introspection, depth, batching",
      parameters: { url: target.url },
      priority: 1,
    })
  }

  // Upload testing
  if (target.hasUpload) {
    vulnSteps.push({
      tool: "upload_test",
      description: "Test file upload bypass",
      parameters: { url: target.url },
      priority: 2,
    })
  }

  // Race conditions (standard/deep only)
  if (scope !== "quick") {
    vulnSteps.push({
      tool: "race_test",
      description: "Test for race conditions on state-changing endpoints",
      parameters: { url: target.url },
      priority: 3,
    })
  }

  // Technology-specific tests
  const techs = target.technologies ?? []
  if (techs.some((t) => /flask|jinja|django/i.test(t))) {
    notes.push("Python web framework detected — prioritize SSTI and deserialization testing")
  }
  if (techs.some((t) => /node|express/i.test(t))) {
    notes.push("Node.js detected — test for prototype pollution, NoSQL injection")
  }
  if (techs.some((t) => /java|spring/i.test(t))) {
    notes.push("Java detected — test for deserialization, XXE, Log4Shell")
  }
  if (techs.some((t) => /php/i.test(t))) {
    notes.push("PHP detected — prioritize LFI, type juggling, deserialization")
  }

  phases.push({ phase: "vuln_testing", steps: vulnSteps })

  // Phase 4: Exploitation Validation (standard/deep only)
  if (scope !== "quick") {
    phases.push({
      phase: "exploitation",
      steps: [
        {
          tool: "build_chains",
          description: "Build attack chains from discovered findings",
          parameters: {},
          priority: 1,
        },
      ],
    })
  }

  // Phase 5: Reporting
  phases.push({
    phase: "reporting",
    steps: [
      {
        tool: "generate_report",
        description: "Generate assessment report",
        parameters: { format: "markdown" },
        priority: 1,
      },
    ],
  })

  return {
    target: target.url,
    scope,
    phases,
    timeout: timeouts[scope],
    notes,
  }
}

// Replan signals — conditions that should modify the plan
export interface ReplanSignal {
  type:
    | "waf_detected"
    | "auth_obtained"
    | "technology_identified"
    | "rate_limited"
    | "escalation_found"
    | "large_surface"
    | "api_detected"
    | "spa_detected"
    | "graphql_detected"
  data?: Record<string, any>
}

export function applyReplanSignal(plan: AttackPlan, signal: ReplanSignal): AttackPlan {
  const updated = { ...plan, phases: plan.phases.map((p) => ({ ...p, steps: [...p.steps] })) }

  switch (signal.type) {
    case "waf_detected":
      updated.notes.push("WAF detected — use encoding/evasion techniques in payloads")
      break

    case "auth_obtained":
      updated.notes.push("Credentials obtained — re-test all endpoints with authentication")
      break

    case "rate_limited":
      updated.notes.push("Rate limiting detected — reduce concurrency, add delays")
      break

    case "graphql_detected": {
      const vulnPhase = updated.phases.find((p) => p.phase === "vuln_testing")
      if (vulnPhase && !vulnPhase.steps.some((s) => s.tool === "graphql_test")) {
        vulnPhase.steps.push({
          tool: "graphql_test",
          description: "GraphQL endpoint discovered — test for introspection, depth, batching",
          parameters: { url: signal.data?.url ?? plan.target },
          priority: 1,
        })
      }
      break
    }

    case "large_surface":
      updated.notes.push("Large attack surface — focus on highest-priority tests, parallelize where possible")
      break

    case "escalation_found":
      updated.notes.push("Privilege escalation opportunity — validate and chain with existing findings")
      break

    default:
      break
  }

  return updated
}

export function formatPlan(plan: AttackPlan): string {
  const lines: string[] = [
    `── Attack Plan ──`,
    `Target: ${plan.target}`,
    `Scope: ${plan.scope}`,
    `Timeout: ${plan.timeout}s`,
    "",
  ]

  for (const phase of plan.phases) {
    lines.push(`Phase: ${phase.phase.toUpperCase()}`)
    for (const step of phase.steps) {
      const pri = step.priority === 1 ? "MUST" : step.priority === 2 ? "SHOULD" : "MAY"
      lines.push(`  [${pri}] ${step.tool}: ${step.description}`)
    }
    lines.push("")
  }

  if (plan.notes.length > 0) {
    lines.push("Notes:")
    for (const note of plan.notes) lines.push(`  • ${note}`)
  }

  return lines.join("\n")
}

export {
  createPlannerKernel,
  applyPlannerEvent,
  nextPlannerStep,
  plannerIsTerminal,
  type PlannerKernel,
  type PlannerEvent,
  type PlannerState,
} from "./kernel"
